### PR TITLE
Initial support for X-Forward-For 

### PR DESCRIPTION
**Pull request by mths1**
Initial support for X-Forward-For  header. Currently, only works for Websockets (MQTTWS). To turn on support for X-Forward-Header you need to set

- listener.ws.$name.proxy_xff_support on
- istener.ws.$name.proxy_xff_trusted_intermediate IP1;IP2;...

where IP1... are acceptable proxy IPs. Support for vmq_http_pub plugin will follow (amazing how many changes this small plugin triggers lately) :-)

Related Issues:
#1754 
#1783 
(among others)

This is WiP, as I do not have a test lab for all kinds of proxies any help (testers) is appreciated. The code will print some debug information that will help in case something breaks. 

## Proposed Changes

Adds support for X-Forward-For header.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes issue #XXXX)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, styles...)
- [ ] DevOps (Build scripts, pipelines...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [ ] I have read the [CODE_OF_CONDUCT.md](https://github.com/vernemq/vernemq/blob/main/CODE_OF_CONDUCT.md) document
- [ ] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if needed)
- [ ] Any dependent changes have been merged and published in related repositories
- [ ] I have updated changelog (At the bottom of the release version)
- [ ] I have squashed all my commits into one before merging

## Further Comments

If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc.

### CODE REVIEW

**Comment by mths1** [comment]
How to proceed with this one? I have the feeling letting it in the WiP state will not help :-) It works for my nginx use case, but to get this more stable it needs to be tested with more setups.

**Comment by ioolkos** [comment]
@mths1 please leave it as Draft for now, so that I can test at least HAProxy. (after that we can take out the test info logs).
Also, even if it turns out to be only useful/working on NGINX, it'd still be a valuable contribution, precisely because NGINX is missing a full proxy protocol implementation at the outgoing side.

**Comment by mths1** [comment]
@ioolkos : Any news on this? I'd have some time this week to "get stuff done".

**Comment by ioolkos** [comment]
@mths1 ah, yes, I promised to check this on HAProxy... need to find the hour to do this.
Other than that, the PR seems to look complete, so feel free to promote it from draft status.

**Comment by ioolkos** [comment]
@mths1 on a first quick test this doesn't seem to work on HAProxy yet.

Can you show the exact client side request you use for NGINX?

**Comment by ioolkos** [comment]
Ok, I can establish a WS connection over HAProxy, with a config as below. Haven't yet checked further details.

in `haproxy.cfg`:

```
frontend vernemq_ws_frontend_xff
        bind 127.0.0.1:9113 name http1
	mode http
        default_backend vernemq_ws_backend_xff
        
backend vernemq_ws_backend_xff
	option http-server-close
	option forwardfor
	mode http
   	timeout tunnel 1h
        server mqtt1 127.0.0.1:7779 check
```

in `vernemq.conf`:
```
listener.ws.xfftest = 127.0.0.1:7779
listener.ws.xfftest.proxy_xff_support = on
listener.ws.xfftest.proxy_xff_trusted_intermediate = 127.0.0.1
```

**Comment by mths1** [comment]
@ioolkos : Great! My nginx test was quite similar (and thus quite basic). I see some reproducible error in the vmq_connect_suite which I will look into.

**Comment by ioolkos** [comment]
@mths1 You mentioned an error in the connect_suite. Were you able to resolve that? if yes, I think in terms of the feature, we can go with what we have here.

File: apps/vmq_server/src/vmq_ranch_config.erl [hunk 1]
```diff
@@ -299,6 +299,7 @@ protocol_opts(vmq_ranch, _, Opts) ->
 protocol_opts(cowboy_clear, Type, Opts) when
     (Type == mqttws) or (Type == mqttwss)
 ->
+    lager:info("My opts ~p~n", [Opts]),
```
Comment (ioolkos) [inline apps/vmq_server/src/vmq_ranch_config.erl:302]: to remove

File: apps/vmq_server/src/vmq_websocket.erl [hunk 2]
```diff
@@ -46,6 +46,7 @@
 
 init(Req, Opts) ->
     Type = proplists:get_value(type, Opts),
+    lager:info("Options ~p~n", [Opts]),
```
Comment (ioolkos) [inline apps/vmq_server/src/vmq_websocket.erl:49]: to remove

**Review by ioolkos — CHANGES_REQUESTED** [review]
A couple logs to remove, then we merge the feature.

**Pull request merged**